### DISCUSSION TITLE
Questions about Fault-tolerant Training

### DISCUSSION THREAD

**Discussion by Wuziyi616** [Lightning-AI/pytorch-lightning#10380] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi! I'm working on a SLURM cluster with preemption, so I'm really excited to see the support of Fault-tolerant Training in 1.5.0. However, when I upgrade package and try `PL_FAULT_TOLERANT_TRAINING=1 python train.py xxx` in the cluster, it doesn't seem to work.

I look into the code of `Trainer`, it seems that the code responsible for fault-tolerant is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/45f6a3b1758f88af7fd776915539800cbc0137a9/pytorch_lightning/trainer/trainer.py#L668). I assume preemption is a `BaseException` so the code will go to [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/45f6a3b1758f88af7fd776915539800cbc0137a9/pytorch_lightning/trainer/trainer.py#L693) and finally [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/45f6a3b1758f88af7fd776915539800cbc0137a9/pytorch_lightning/trainer/trainer.py#L1550) so that we save a checkpoint?

However, when set some print in the code, when I use ctrl+C to interrupt code, it indeed goes to [this](https://github.com/PyTorchLightning/pytorch-lightning/blob/45f6a3b1758f88af7fd776915539800cbc0137a9/pytorch_lightning/trainer/trainer.py#L681) `KeyBoardInterrupt`. But if I use `scontrol requeue` to simulate a preemption, the code didn't got to `BaseException`. And that's why it didn't save a checkpoint for Fault-tolerant Training.

Is there anything wrong with my code? I assume interruptions like `scancel` `requeue` are considered in this case. Can anyone help me? Thank you in advance!

EDIT: I've looked in the code a little bit more, it seems that when I do `scancel` or `scontrol requeue`, the code directly exit, without throwing an exception, and that's why it didn't go to the `except  _on_exception` section. Is this expected behavior? Or is there anyway to solve it?

I think that's related to the signal that SLURM sent to my program, and I already see a `SignalConnector` dealing with SLURM in pytorch-lightning [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/45f6a3b1758f88af7fd776915539800cbc0137a9/pytorch_lightning/trainer/connectors/signal_connector.py#L40). I also see [this answer](https://stackoverflow.com/questions/65326039/how-to-handle-job-cancelation-in-slurm) about the signal of SLURM. Maybe I should set it in the sbatch script? Any suggestions?

**Comment by Wuziyi616** [comment, marked as answer]
Solved. That's indeed because in my SLURM cluster, there is no time interval between signal sending and program killing, so PyTorch-Lightning just don't have time to do checkpointing

**Reply by tchaton** [reply]
I don't have access to a SLURM cluster, so I can't test it out, but in theory, if you use `PL_FAULT_TOLERANT_TRAINING=1`, the previous behaviour with [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/signal_connector.py#L52) should work as expected.
